File: adc03.py
import re

digit_reg = re.compile("\\d+")
symbol_reg = re.compile("[^\\.\\d]")


numbers = {}


def get_digits(ln, str):
    for m in digit_reg.finditer(str):
        numbers[(ln, m.span())] = int(m.group())


symbols = {}


def get_symbols(ln, str):
    for m in symbol_reg.finditer(str):
        symbols[(ln, m.start())] = m.group()


def surround(ln, a, b):
    lst = []

    def test(x, y):
        if (x, y) in symbols:
            lst.append((x, y))

    test(ln, a - 1)
    test(ln, b)
    for c in range(a - 1, b + 1):
        test(ln - 1, c)
        test(ln + 1, c)
    return lst


with open("input03.txt", "r") as file:
    for ln, line in enumerate(file):
        line = line[:-1]
        get_digits(ln, line)
        get_symbols(ln, line)
    sum1 = 0
    for k in numbers:
        if surround(k[0], k[1][0], k[1][1]):
            sum1 += numbers[k]
    print("Part one:", sum1)
    sum2 = 0
    gears = {k: [] for k in symbols if symbols[k] == "*"}
    for k in numbers:
        for s in surround(k[0], k[1][0], k[1][1]):
            if s in gears:
                gears[s].append(numbers[k])
    sum2 = 0
    for g in gears:
        if len(gears[g]) == 2:
            sum2 += gears[g][0] * gears[g][1]
    print("Part two:", sum2)

File: adc03/adc03.ml
(* let file = "test03.txt" *)
let file = "input03.txt"
let is_symbol = function '.' | '0' .. '9' -> false | _ -> true

let is_digit = function
  | '0' .. '9' as c -> Some (Char.code c - 48)
  | _ -> None

let read_grid data =
  let lines = ref [] in
  (try
     while true do
       lines :=
         (input_line data |> String.to_seq |> Array.of_seq)
         :: !lines
     done
   with End_of_file -> ());
  !lines |> List.rev |> List.to_seq |> Array.of_seq

let get_surrounding tbl l a b =
  let srnd = ref [] in
  let test k = if Hashtbl.mem tbl k then srnd := k :: !srnd in
  test (l, a - 1);
  test (l, b);
  for c = a - 1 to b do
    test (l - 1, c);
    test (l + 1, c)
  done;
  !srnd

let get_symbols grid =
  let symbols = Hashtbl.create 10 in
  for l = 0 to Array.length grid - 1 do
    for c = 0 to Array.length grid.(0) - 1 do
      let char = grid.(l).(c) in
      if is_symbol char then Hashtbl.add symbols (l, c) char
    done
  done;
  symbols

let get_numbers grid =
  let numbers = Hashtbl.create 10 in
  let rec scan_line line col =
    if col < Array.length grid.(line) then
      match is_digit grid.(line).(col) with
      | Some d -> scan_number line (col + 1) col d
      | None -> scan_line line (col + 1)
  and scan_number line col start number =
    if col < Array.length grid.(line) then (
      match is_digit grid.(line).(col) with
      | Some d ->
          scan_number line (col + 1) start ((10 * number) + d)
      | None ->
          Hashtbl.add numbers (line, start, col) number;
          scan_line line (col + 1))
    else Hashtbl.add numbers (line, start, col) number
  in
  for line = 0 to Array.length grid - 1 do
    scan_line line 0
  done;
  numbers

let _ =
  let data = open_in file in
  let grid = read_grid data in
  let numbers = get_numbers grid
  and symbols = get_symbols grid in
  Printf.printf "Part one: %d\n"
    (Hashtbl.fold
       (fun (l, a, b) n sum ->
         match get_surrounding symbols l a b with
         | [] -> sum
         | _ -> sum + n)
       numbers 0);
  let gears = Hashtbl.create 10 in
  Hashtbl.iter
    (fun pos symb -> if symb = '*' then Hashtbl.add gears pos [])
    symbols;
  Hashtbl.iter
    (fun (l, a, b) n ->
      List.iter
        (fun pos ->
          Hashtbl.replace gears pos (n :: Hashtbl.find gears pos))
        (get_surrounding gears l a b))
    numbers;
  Printf.printf "Part two: %d\n"
    (Hashtbl.fold
       (fun _ l sum ->
         match l with [ a; b ] -> sum + (a * b) | _ -> sum)
       gears 0)
